src/codec.ts

```diff
@@ -39,22 +39,306 @@ export interface JSONOptions {
 }
 
 /**
- * @typedef SpannerDate
+ * Date-like object used to represent Cloud Spanner Dates.
+ *
  * @see Spanner.date
+ * @see https://cloud.google.com/spanner/docs/data-types#date-type
+ *
+ * @class
+ * @extends Date
+ *
+ * @param {string|number} [date] String or number representing the date.
+ *
+ * @example
+ * Spanner.date('3-3-1933');
  */
-export class SpannerDate {
-  value: string;
-  constructor(value?: string|number|Date) {
-    if (arguments.length > 1) {
-      throw new TypeError([
-        'The spanner.date function accepts a Date object or a',
-        'single argument parseable by Date\'s constructor.',
-      ].join(' '));
+export class SpannerDate extends Date {
+  constructor(value?: string|number) {
+    const date = new Date(value || Date.now());
+    const dateStr = SpannerDate.getDateString(date);
+
+    super(dateStr);
+  }
+  /**
+   * Returns the date in Cloud Spanner date format.
+   * `YYYY-[M]M-[D]D`
+   *
+   * @returns {string}
+   */
+  toJSON() {
+    return SpannerDate.getDateString(this);
+  }
+  /**
+   * Formats an ISO string into a Cloud Spanner Date string.
+   *
+   * @private
+   *
+   * @param {Date} date The date to extract the `YYYY-[M]M-[D]D` value from.
+   */
+  static getDateString(date: Date): string {
+    return date.toISOString().replace(/T.+/, '');
+  }
+}
+
+/**
+ * Date-like object used to represent Cloud Spanner Timestamps.
+ *
+ * @see Spanner.timestamp
+ * @see https://cloud.google.com/spanner/docs/data-types#timestamp-type
+ *
+ * @class
+ * @extends Date
+ *
+ * @example
+ * const timestamp = Spanner.timestamp('2019-02-08T10:34:29.481145231Z');
+ *
+ * @example <caption>With a `google.protobuf.Timestamp` object</caption>
```

Comment:
Should we just call this "an object" instead of the google.protobuf thing? I'm not sure (m)any users are touching that. Or, if I'm wrong, we could link it to the definition of that type, so users can see what options they have. But it kind of just feels like a regular old object that we should notate in the params.

Reply:
Thats a good point, I was sort of wondering if we should instead accept a tuple to mirror whats going on with [`process.hrtime()`](https://nodejs.org/api/process.html#process_process_hrtime_time)